src/vm/amd64/cgenamd64.cpp

```diff
@@ -1126,6 +1126,121 @@ PCODE DynamicHelpers::CreateHelperWithTwoArgs(LoaderAllocator * pAllocator, TADD
+// Assumption: the first argument is already placed into the first argument register.
+//
+void DynamicHelpers::EmitJumpStubWithSecondArg(LoaderAllocator * pAllocator, BYTE*& pCode, PCODE pTarget, TADDR secondArgValue)
+{
+    CONTRACTL
+    {
+        GC_NOTRIGGER;
+        PRECONDITION(pCode != NULL && pTarget != NULL);
+    }
+    CONTRACTL_END;
+
+    // Move second argument into argument register (size = 10)
+#ifdef UNIX_AMD64_ABI
+    *(UINT16 *)pCode = 0xBE48; // mov rsi, XXXXXX
+#else
+    *(UINT16 *)pCode = 0xBA48; // mov rdx, XXXXXX
+#endif
+    pCode += 2;
+    *(TADDR *)pCode = secondArgValue;
+    pCode += 8;
+
+    // Jump stub (size = 5)
+    *pCode++ = X86_INSTR_JMP_REL32; // jmp rel32
+    *(INT32 *)pCode = rel32UsingJumpStub((INT32 *)pCode, pTarget, NULL, pAllocator);
+    pCode += 4;
+}
+
+PCODE DynamicHelpers::CreateDictionaryLookupHelper(LoaderAllocator * pAllocator, CORINFO_RUNTIME_LOOKUP * pLookup)
+{
+    STANDARD_VM_CONTRACT;
+
+    // TODO: fix codegen for correct Unix ABI...
+
+    // It's available only via the run-time helper function
+    if (pLookup->indirections == CORINFO_USEHELPER)
+    {
+        BEGIN_DYNAMIC_HELPER_EMIT(16);
+
+        // rcx contains the generic context parameter
+        // mov rdx,pLookup->signature
+        // call pLookup->helper
+        EmitJumpStubWithSecondArg(pAllocator, p, CEEJitInfo::getHelperFtnStatic(pLookup->helper), (TADDR)pLookup->signature); // 15 bytes
+
+        // ret
+        *p++ = 0xC3;
+
+        END_DYNAMIC_HELPER_EMIT();
+    }
+    else
+    {
+        int codeSize = (pLookup->indirections * 4) + (pLookup->testForNull ? 28 : 4);
+
+        BEGIN_DYNAMIC_HELPER_EMIT(codeSize);
+
+        if (pLookup->testForNull)
+        {
+            // rcx contains the generic context parameter. Push on the stack to save a copy of it.
+            *p++ = 0x51;    // push rcx
+        }
+
+        for (WORD i = 0; i < pLookup->indirections; i++)
+        {
+            _ASSERTE(pLookup->offsets[i] >= 0 && pLookup->offsets[i] < 256);
+
+            // mov rcx,qword ptr [rcx+offset]
+            *(UINT32*)p = 0x00498b48; p += 3;
+            *p++ = (BYTE)pLookup->offsets[i];
+        }
+
+        // No null test required
+        if (!pLookup->testForNull)
+        {
+            // No fixups needed for R2R
+
+            // mov rax,rcx
+            *(UINT32*)p = 0x00c88948; p += 3;
+            *p++ = 0xC3;    // ret
+        }
+        else
+        {
+            // rcx contains the value of the dictionary slot entry
+
+            _ASSERTE(pLookup->indirections != 0);
+
+            // test rcx,rcx
+            *(UINT32*)p = 0x00c98548; p += 3;
+
+            // jne 'POP_AND_RETURN'
+            *(UINT16*)p = 0x1175; p += 2;
+
+            // pop rcx
+            *p++ = 0x59;        // pop the generic context back into rcx
+
+            // mov rdx,pLookup->signature
+            // call pLookup->helper
```

Comment:
Resolved

---
In reply to: [60820448](https://github.com/dotnet/coreclr/pull/4519#discussion_r60820448) [](ancestors = 60820448)